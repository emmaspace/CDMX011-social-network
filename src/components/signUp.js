/* eslint-disable import/no-cycle */
/* eslint-disable no-param-reassign */
/* eslint-disable quotes */
import { signUpWithPassword, userProfile } from "../lib/firebase.js";
import { onNavigate } from "../routes.js";

export const signUp = (target) => {
  const signUpContainer = `
    <header>
        <img src="./assets/logo.png" alt="Logo de Nova" id="img-signup"/>
    </header>
    <main>
    <form class= "signUp color-line" id="formRegister">
      <div class = "formContainer">
        <label for="email">e-mail*</label><br>
        <input type="email" id="email" required><br>
        <label for="user">Nombre de usuario*</label><br>
        <input type="text" id="user" required><br>
        <label for="password">Contraseña*</label>
        <span id="show"><span class="iconify" data-icon="emojione-monotone:eye" style="color: white;" data-height="26"></span></span>       
        <span id="hide"><span class="iconify" data-icon="akar-icons:eye-closed" style="color: white;" data-height="26"></span></span><br>
        <input type="password" id="password" required><br>
        <label for="passwordRepeat">Repite tu contraseña*</label>
        <span id="showRepeat"><span class="iconify" data-icon="emojione-monotone:eye" style="color: white;" data-height="26"></span></span>       
        <span id="hideRepeat"><span class="iconify" data-icon="akar-icons:eye-closed" style="color: white;" data-height="26"></span></span><br>
        <input type="password" id="passwordRepeat" required><br>  
      </div>
      <p id="messageError"></p>
      <button class="signUpButton" type ="submit" id = "signUpButton"> Registrarse </button>
    </form>
    </main>    
    <footer class= "signUpFooter">
      <a href="#" id="back" aria-label="Link para regresar"><img src="./assets/back.png" alt="Ícono para regresar"></a>
    </footer>
`;

  target.innerHTML = signUpContainer;

  const toLogIn = document.getElementById("back");
  toLogIn.addEventListener("click", (event) => {
    event.preventDefault();
    onNavigate("/");
  });

  const signUpBttn = document.getElementById("signUpButton");
  signUpBttn.addEventListener("click", (event) => {
    event.preventDefault();
    const email = document.getElementById("email").value;
    const password = document.getElementById("password").value;
    const passwordRepeat = document.getElementById("passwordRepeat").value;
    let user = document.getElementById("user").value;
    if (password === passwordRepeat) {
      if (user !== "") {
        signUpWithPassword(email, password)
          .then(() => {
            // document.getElementById("formRegister").reset();
            user = document.getElementById("user").value;
            userProfile(user);
            onNavigate("/home");
          })
          .catch((error) => {
            document.getElementById("messageError").innerText = error.message;
          });
      } else {
        document.getElementById("messageError").innerText = "Ingresa un nombre de usuario";
      }
    } else {
      document.getElementById("messageError").innerText = "Las contraseñas no coinciden";
    }
  });

  const password = document.getElementById('password');
  const showButton = document.getElementById('show');
  const hideButton = document.getElementById('hide');

  const showPassword = () => {
    if (password.type === 'password') {
      password.setAttribute('type', 'text');
      showButton.style.display = 'none';
      hideButton.style.display = 'inline-block';
    } else if (password.type === 'text') {
      showButton.style.display = 'inline-block';
      hideButton.style.display = 'none';
      password.setAttribute('type', 'password');
    }
  };
  showButton.addEventListener('click', showPassword);
  hideButton.addEventListener('click', showPassword);

  const passwordRepeat = document.getElementById('passwordRepeat');
  const showButtonRepeat = document.getElementById('showRepeat');
  const hideButtonRepeat = document.getElementById('hideRepeat');

  const showPasswordRepeat = () => {
    if (passwordRepeat.type === 'password') {
      passwordRepeat.setAttribute('type', 'text');
      showButtonRepeat.style.display = 'none';
      hideButtonRepeat.style.display = 'inline-block';
    } else if (passwordRepeat.type === 'text') {
      showButtonRepeat.style.display = 'inline-block';
      hideButtonRepeat.style.display = 'none';
      passwordRepeat.setAttribute('type', 'password');
    }
  };
  showButtonRepeat.addEventListener('click', showPasswordRepeat);
  hideButtonRepeat.addEventListener('click', showPasswordRepeat);
};
