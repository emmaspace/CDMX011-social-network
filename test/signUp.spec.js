/**
 * @jest-environment jsdom
 */
/* eslint-disable max-len */
// importamos la funcion que vamos a testear
// import { onNavigate } from '../src/app.js';
import './globals/firebase-test.js';
import { signUp } from '../src/components/signUp.js';

const delay = (timeInMs = 100) => new Promise((resolve) => setTimeout(resolve, timeInMs));

describe('signUp', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>';
  });
  const signUpForm = (email, password) => {
    document.getElementById('email').value = email;
    document.getElementById('password').value = password;
    document.getElementById('signUpButton').click();
  };

  it('debería ser una función', () => {
    expect(typeof signUp).toBe('function');
  });
  // eslint-disable-next-line jest/no-focused-tests
  it('debería desplegar la pantalla de signUp', () => {
    const rootDiv = document.getElementById('root');
    signUp(rootDiv);
    expect(rootDiv.innerHTML).toMatchSnapshot();
  });

  it('registro del usuario exitoso', async () => {
    const rootDiv = document.getElementById('root');

    const email = 'test@test.com';
    const password = '123456';
    const userConsole = { email, uid: 'xxxxxxx' };

    const mockCreateUserWithEmailAndPassword = jest.fn();
    mockCreateUserWithEmailAndPassword.mockResolvedValue(userConsole);

    const mockFirebaseAuth = {
      createUserWithEmailAndPassword: mockCreateUserWithEmailAndPassword,
      currentUser: userConsole,
    };

    firebase.auth = () => mockFirebaseAuth;

    signUp(rootDiv);
    signUpForm(email, password);

    await delay(100);

    expect(mockCreateUserWithEmailAndPassword).toHaveBeenCalledWith(email, password);
  });
});
