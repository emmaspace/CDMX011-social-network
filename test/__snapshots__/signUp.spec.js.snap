// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signUp debería desplegar la pantalla de signUp 1`] = `
"
    <header>
        <img src=\\"./assets/logo.png\\" alt=\\"Logo de Nova\\" id=\\"img-signup\\">
    </header>
    <main>
    <form class=\\"signUp color-line\\" id=\\"formRegister\\">
      <div class=\\"formContainer\\">
        <label for=\\"email\\">e-mail*</label><br>
        <input type=\\"email\\" id=\\"email\\" required=\\"\\"><br>
        <label for=\\"user\\">Nombre de usuario*</label><br>
        <input type=\\"text\\" id=\\"user\\" required=\\"\\"><br>
        <label for=\\"password\\">Contraseña*</label><br>
        <input type=\\"password\\" id=\\"password\\" required=\\"\\"><br>
        <label for=\\"passwordRepeat\\">Repite tu contraseña*</label><br>
        <input type=\\"password\\" id=\\"passwordRepeat\\" required=\\"\\"><br>  
      </div>
      <p id=\\"messageError\\"></p>
      <button class=\\"signUpButton\\" type=\\"submit\\" id=\\"signUpButton\\"> Registrarse </button>
    </form>
    </main>    
    <footer class=\\"signUpFooter\\">
      <a href=\\"#\\" id=\\"back\\" aria-label=\\"Link para regresar\\"><img src=\\"./assets/back.png\\" alt=\\"Ícono para regresar\\"></a>
    </footer>
"
`;
